lib/fog/openstack/core.rb

```diff
@@ -119,6 +116,56 @@ def set_api_path
         # if the service supports multiple versions, do the selection here
       end
 
+      def set_microversion
+        # both need to be set in service's initialize for microversions to work
+        return if !@supported_microversion || !@supported_versions
+
+        @microversion_key          ||= 'Openstack-API-Version'.freeze
+        @microversion_service_type ||= @openstack_service_type.first
+
+        @microversion = Fog::OpenStack.get_supported_microversion(
+          @supported_versions,
+          @openstack_management_uri,
+          @auth_token,
+          @connection_options
+        ).to_s
+
+        # choose minimum out of reported and supported version
+        if microversion_newer_than?(@supported_microversion)
+          @microversion = @supported_microversion
+        end
+
+        # choose minimum out of set and wished version
+        if @fixed_microversion && microversion_newer_than?(@fixed_microversion)
+          @microversion = @fixed_microversion
+        elsif @fixed_microversion && @microversion != @fixed_microversion
+          Fog::Logger.warning("Microversion #{@fixed_microversion} not supported")
+        end
+      end
+
+      def microversion_newer_than?(version)
+        Gem::Version.new(version) < Gem::Version.new(@microversion)
```

Comment:
Also using Gem `version` this way works, I don't think that's the standard approach.
Effectively it escapes Gem dependency (handled through gem specs). That would also need a `require` statement.